### PR TITLE
Feature/implement dask

### PR DESCRIPTION
**Pull request by mrijken**
Do not merge yet; work in progress.

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -4,7 +4,6 @@ cover_pylib = True
 
 [report]
 show_missing = True
-fail_under = 60
```
[the remark this answers is not in the document]
Comment (mrijken) [reply .coveragerc]: Done

File: README.md [hunk 2]
```diff
@@ -139,6 +139,12 @@ The unittests for the project can be run using [pytest](https://pytest.org/):
 pytest
 ```
 
+At default, the pandas implementation will be used for the test. You can use test the Dask implementation via:
```
[the remark this answers is not in the document]
Comment (mrijken) [reply README.md]: Both are tested. Docs where nor updated. Done.

File: README.md [hunk 3]
```diff
@@ -139,6 +139,8 @@ The unittests for the project can be run using [pytest](https://pytest.org/):
 pytest
 ```
 
+All tests are run with both the Pandas and Dask engine.
```
Comment (wpbs) [inline README.md]:
```suggestion
All tests are executed with both the Pandas and Dask engine.
```

File: dvb/datascience/eda/boxplot.py [hunk 4]
```diff
@@ -19,17 +23,105 @@ class BoxPlot(PipeBase):
...
         df = data["df"]
 
         display(HTML("<h2>Boxplots Transform %s</h2>" % params["metadata"]["name"]))
 
         for feature in df.columns:
             fig = self.get_fig((params["metadata"]["name"], feature))
-            plt.boxplot([df[feature]])
+
+            self._plot_boxplot(self._describe_column(df[feature]))
+            # plt.boxplot([df[feature]])
```
Comment (wpbs) [inline dvb/datascience/eda/boxplot.py]: commented code should be removed

File: dvb/datascience/eda/dimension_reduction.py [hunk 5]
```diff
@@ -79,6 +79,9 @@ def transform(self, data: Data, params: Params) -> Data:
 
         return {"figs": self.figs}
 
+    def transform_dask(self, data: Data, params: Params) -> Data:
+        raise NotImplementedError
```
Comment (wpbs) [inline dvb/datascience/eda/dimension_reduction.py]: Perhaps a good idea to document this (missing) feature somewhere (comments over here / action item on the backlog?)

File: dvb/datascience/pipeline.py [hunk 6]
```diff
@@ -573,6 +596,7 @@ def transform(
                 if fit:
                     # fit_and_transform
                     logger.info("fit pipe %s", pipe.name)
+                    print
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]: why print nothing here?

File: dvb/datascience/pipeline.py [hunk 7]
```diff
@@ -425,6 +432,20 @@ def get_pipe_input(self, name) -> Optional[Dict]:
...
 
+    @classmethod
+    def _compute_when_needed(cls, data: Data):
+        """
+        Data can contain lazy DataFrames. When a df is lazy, that one is computed.
+        """
+        d = {}
+
+        for key, df in data.items():
+            if type(df) in (dd.DataFrame, dd.Series):
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]: check using isinstance() instead of type(df)

File: dvb/datascience/transform/impute.py [hunk 8]
```diff
@@ -120,36 +157,79 @@ def _get_mask(X, value):
...
-        for column in df.columns:
-            mask = self._get_mask(df[column], self.missing_values)
-            df[column][mask] = self.fill[column]
+        features = self.features or data["df"].columns
+        for feature in features:
+            df[feature] = df[feature].apply(lambda value: self.fill[feature] if value in self.missing_values else value)
 
         return {"df": df}
+
+    # def transform_dask(self, data: Data, params: Params) -> Data:
```
Comment (wpbs) [inline dvb/datascience/transform/impute.py]: commented code can be removed

File: setup.py [hunk 9]
```diff
@@ -43,16 +43,20 @@
         "blockdiag",
         "pyensae",
         "TPOT",
-    ],  # Optional
-    extras_require={  # Optional
+        "xlrd",
+        # TODO: make dask optional
```
Comment (wpbs) [inline setup.py]: fix TODO

Comment (mrijken) [reply setup.py]: TODO won't be done; impact is too heavy.

File: README.md [hunk 10]
```diff
@@ -131,6 +131,13 @@ This example shows:
 
 For a more extensive overview of all the features, see the docs directory.
 
+## Backends
+
+Both Pandas and Dask can be used for the actual processing. At default Pandas will be used. When needed, Dask can be
+used to handle datasets which do not fit in memory or when distributed computing is needed. Dask does is not a drop-in
```
Comment (wpbs) [inline README.md]: does is -> is

**Pull request merged**